merge.go

```diff
@@ -114,20 +119,41 @@ next:
 	return out
 }
 
-// lookupByHeader optionally returns a File from fs.files if the FileHeaders match.
+// findOutfile optionally returns a File from fs.files if the FileHeaders match.
 // This is done because we append batches into files to minimize the count of output files.
 //
-// lookupByHeader will return the existing file (stored in outfiles) if no matching file exists.
-func (fs *mergableFiles) lookupByHeader(f *File) *File {
-	for i := range fs.outfiles {
-		if fs.outfiles[i].Header.ImmediateDestination == f.Header.ImmediateDestination &&
-			fs.outfiles[i].Header.ImmediateOrigin == f.Header.ImmediateOrigin {
-			// found a matching file, so return it
-			return fs.outfiles[i]
+// findOutfile will return the existing file (stored in outfiles) if no matching file exists.
+func (fs *mergableFiles) findOutfile(f *File) *File {
+	var lookup func(int) *File
+	lookup = func(start int) *File {
```

Comment:
Could you add some additional code comments about how this recursive function does the lookup? Like a high-level overview